stereograph.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" $Id: stereograph.1,v 1.6 2003/10/19 11:46:10 janoschs Exp $
.TH STEREOGRAPH 1 "October, 2003"
.\" Please adjust this date whenever revising the manpage.

.SH NAME
stereograph \- stereogram generator


.SH SYNOPSIS
.LP
\fBstereograph\fP [\fIoptions\fP] -b \fIbase\fP [-t \fItexture\fP] [-o \fIoutput\fP]


.SH DESCRIPTION

Stereograph is a stereogram generator. In detail it is a single image
stereogram (SIS) generator. It produces twodimensional images that
seem to be threedimensional.

\fIbase\fP is the file describung the \*(rqrelief\*(lq of the 3d
model. It is a simple graphic where the brightness of a pixel defines
its individual depth.  The darker a pixel is, the more far away it
will seem to be in the final stereogram - the brighter it is, the
closer it will be.
.P
The \fItexture\fP is that what everyone sees when regarding your stereogram
\- even if he cannot get in the third dimensions of the composition.
If you use the \fB-w\fP swich this becomes optional.
.P
The width of the texture stands for the maximum depth of any steregram
and it cannot be greater than the distance of a viewer's two eyes -
otherwise they won't be able to see anything in the stereogram but the
beautiful great texture. As a hand rule, 100 should work nice for
stereograms of 640*480 up to 800*600 pixels. Use 110 to 120 for
bigger ones.
.P
The files given as \fIbase\fP and \fItexture\fP must be supplied
either in the jpg, png, ppm or tga format. Which you used is derived from
the file name.
.P
\fIoutput\fP is the file where the stereogram is written to. The file
format to use is deduced from the filename given if not overridden by
\fB-f\fP. (\*rq.jpg\*(lq, \*rq.png\*(lq, (\*rq.ppm\*(lq, and (\*rq.tga\*(lq 
are supported. If not given, \fBstereograph\fP will write the image in PPM
or the format given with \fB-f\fP format to the standard output.
.P


.SH OPTIONS

.IP "\fB-v\fP"
Be verbose.

.IP "\fB-f\fP \fIoutputformat\fP"
\fIoutputformat\fP is one of the following: jpg, png, ppm or jpg.

.IP "\fB-A\fP"
The Aid flag \fB-A\fP will add a pair of black triangles at the top of
the stereogram to make it easier for unexperienced eyes to achieve the
magic view.

.SS "Quality controling arguments"

.IP "\fB-a\fP \fIanti-aliasing\fP"
\fIanti-aliasing\fP describes a value between 1 and 32 (4 is the
default) that declares how many pixels shall be calculated by the
renderer virtually for ONE pixel.
.IP
You can calculate easily: physically you habe n depth levels, where n
is the width of the texture used for the stereogram. With the AA
feature you now have theoretically a*n levels.  This feature increases
_massively_ the color depth of the output file

.IP "\fB-z\fP \fIzoom\fP"
based on the same idea as AA and has the same effects but physically
increases the file size. Values range from 1 (default) up to 32.

.SS "perspective"

.IP "\fB-d\fP \fIdistance\fP"
\fIdistance\fP describes the distance of your eyes and the virtual
glass that is between you and your stereogram. Allowed are values from
0.0 up to 20.0 where 5.0 is the default.

.IP "\fB-p\fP \fIfront factor\fP"
\fIfront factor\fP defines the finally used depth space; value between
0.0 and 1.0.

.IP "\fB-e\fP \fIeye\-shift\fP"
\fIeye\-shift\fP (a value between -1.0 and 1.0) controls the
perspective along the x axis (left to right). That is if e is positive
the image is shifted slightly to the right. The default value of 0.0
is exactly the centered perspective.


.SS layout

.IP "\fB-x\fP \fItexture\-insert\-x\fP"
\fItexture\-insert\-x\fP is a value between 0 and the width of the
\fItexture\fP controlling where the \fItexture\fP is inserted the
first time and where the rendering process begins its round. default:
0 (left image border).

.IP "\fB-y\fP \fItexture\-insert\-y\fP"
\fItexture\-insert\-y\fP is a value between 0 and the height of the
\fItexture\fP controlling where the \fItexture\fP is inserted the
first time and where the rendering process begins its round. default:
0 (top image border).

.IP "\fB-w\fP \fItexture\-width\fP"
\fItexture\-width\fP specifies the texture width to use for the
stereogram.  This option tells stereograph to generate a random
texture or if a texture was defined it resizes the texture to match
your dimension. Note that this can only reduce the width and
stereograph doesn't care about image ratios. It just cuts your
texture. Random textures cannot be used with transparent rendering.

.IP "\fB-M\fP, \fB-G\fP, \fB-C\fP, \fB-S\fP"
Use one of there flags to define which color type the random texture
should conform to. Use \fB-M\fP for monochrome, \fB-G\fP for a
grayscale or \fB-C\fP for a random color texture. Please note that the
random texture feature cannot be combined with transparent rendering
and that anti-aliasing increases the color depth of you image \- even
if you use a monochrome texture. To disable anti\-aliasing use \fB-a
1\fP.
.IP
The \fB-S\fP flag is experimental \- it generates an artistic random
texture.

.SS "base options"

.IP "\fB-I\fP"
inverts a base (non-transparent rendering only)

.IP "\fB-L\fP"
disables linear rendering.

.IP "\fB-R\fP"
The anti-artefacts feature tells \fBstereograph\fP to process every
second line in reversed order to prevent artefacts. This flag should
be used only with random textures.

.SS "transparancy"
.IP
Transparent stereograms are rendered in the same way as normal
stereograms, the only difference for you is to define more than one
input base file and provide respectively the same number of
textures. The different bases must have the same dimensions, the
textures must equal themselves in the width. You may use the \fB-w\fP
option to let steregraph resize them interactively for you.

.IP "\fB-l\fP (\fBnone\fP|\fBback\fP|\fBtop\fP)"
adjusts the base levels for transparent rendering: the keyword
\fBnone\fP keeps the defined height levels of the base images,
\fBback\fP adjusts the bottom area of all upper layers to the level
defined by the preceding layer and \fBtop\fP adjusts the areas of
upper layers that are not as far as the preceeding layer to the level
of the preceding layer.


.\" .SH "BUGS"


.SH "SEE ALSO"
\fI/usr/share/doc/stereograph/README\fP
.SH AUTHOR
\fBstereograph\fP has been written by Fabian Januszewski
<fabian.linux@januszewski.de>.
.P
This manual page was written by Peter Palfrader
<ppalfrad@cosy.sbg.ac.at> using large parts from the README file, for
the Debian GNU/Linux system (but may be used by others).

.SH "COMMENT"
.P
This manpage documents version 0.32a of \fBstereograph\fP.


ChangeLog
  2000-07-14  * added transparent layer auto sensing which turns
                the -T flag obsolete;
              * added ppm reading;
              * release 0.29a;

  2000-06-21  * file format auto sensing implemented;
              * fixed floating point exception in the
	        random texture generator;
              * general code cleanups;
              * release of 0.28a;

  2000-06-19  * extended random texture generator and added an
                experimental fix for big endidan machines;
              * release 0.27b;

  2000-06-18  * added an experimental artistic random texture
                generator;
              * release of 0.27a;

  2000-06-17  * fixed segmentation fault related to argument list;
              * release of 0.26.1

  2000-06-09  * implemented base level adjustment;
              * added option to the invert base;
              * changed synopsis to define the
                output format (including the constants);
              * release 0.26

  2000-06-06  * fixed right changemap 'black holes';
              * release 0.25.1

  2000-06-03  * added random texture generator
              * reimplemented the anti-artefacts feature
                and changed synopsis (be aware!);
              * some renderer internal fixes;
              * internally shifted some functions;
              * fixed some minor bugs;
              * release 0.25

  2000-06-02  * added the -w argument to define a
                constant texture width, very useful
                for transparent stereograms and free
                texture images;
              * fixed triangle zoom
              * release 0.24

  2000-05-31  * added the -A flag (aid) which will
                add a pair of black triangles to the image
                to help the magic view of a stereogram;
              * release 0.23

  2000-05-29  * fixed starty parametre handling
              * corrected internal renderer statics
              * release 0.22

  2000-05-28  * added linear rendering option
              * changed constants due to GTKstereograph
                which now includes the current renderer
              * release 0.21

  2000-05-23  * fixed segmentation fault in transparent
                memory managment;
              * fixed standard rendering input file problem
              * release 0.20.1a

  2000-05-21  * introduced transparent rendering,
                standard rendering remains untouched;
	      * in contrast to the single layer convention
                for transparency the renderer does NOT
                provide the application with internal gfx
                pointers. that's due to the issue that
                this great amount of memory shall not be
                cleared up with clear_renderer;
              * updated the readme, just for the case :)
	      * release 0.20a

  2000-05-11  * fixed png reading text output if verbose
              * added the -V option to display version
	      * finally added the man page from Peter Palfrader
	      * updated install and uninstall script
	      * release 0.19.1

  2000-04-15  * added file format switch, adjusted arguments
              * release 0.19

  2000-04-14  * implemented ppm output, optional as file
                or to stdout
              * releasing 0.18

  2000-04-13  * implemented png i/o filter
              * release 0.17

  2000-04-12  * finally added the verbosity flag and pipe
                interaction via standard output.
              * release 0.16

  2000-04-10  * added experminental anti-artefacts feature
              * release 0.15

  2000-04-08  * fixed a minor bug in the aa implementation
                including the aa-zoom implementation and
                even zoom
              * 0.14

  2000-03-31  * changed renderer memory management to speed
                up aa and zoom rendering
              * extended the README
              * release 0.13

  2000-02-16  * fixed aa rounding problem in the two aa
                rendering functions where the mem' is
                flushed back
              * setting up 0.12

  2000-02-11  * added eye perspective argument -e
              * releasing version 0.11

  2000-02-08  * releasing version 0.10 into the playground
                of the linux community. :)



stereograph.c
/* Stereograph 0.28a, 21/06/2000;
 * Copyright (c) 2000 by Fabian Januszewski <fabian.linux@januszewski.de>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */


#include <stdio.h>
#include <stdlib.h>

#define STEREOGRAPH_VERSION "0.28a"

#include "renderer.h"
#include "gfxio.h"
#include "stereograph.h"
#include "globals.h"

char *base_file_name = NULL;
char *texture_file_name = NULL;
char *stereo_file_name = NULL;

char **T_base_file_name = NULL;
char **T_texture_file_name = NULL;

int output_format = 0;

/* interface to the renderer */
struct PARAMS *pParam;
struct GFX_DATA *pBase, *pTexture, *pStereo;

struct GFX_DATA ***T_pBase, ***T_pTexture;

int texture_width_to_use = 0;
int random_texture_type = TEX_GRAYSCALE;


void print_info(void) {
	printf("SYNOPSIS\n  stereograph [-T n] [OPTIONS] -b [FILE] [-t [FILE]] [-w n]\n              [-o [FILE]] [-f png/ppm/tga] [-l none/back/top]\n\n");
	printf("OPTIONS\n  -T followed by the number of layers n\n     enables the transparency option\n  -a anti-aliasing     -z zoom               (quality)\n  -d distance          -e eye shift          (perspective)\n  -x texture insert x  -y texture insert y   (layout)\n  -w texture width to use\n  -M/-G/-C generate a monochrome, grayscale or colored\n  -S generates an experimental random texture\n           random texture (no transparency)\n  -A add a pair of triangles                 (aid)\n  -R this flag enables the anti-artefact feature\n  -L disables the linear rendering algorithm\n  -I invert the base\n  -l sets the level adjust mode for transparent rendering\n  -f defines the output format\n  -v you should know this flag :)\n  -V display version\n\n");
}



int main (int argc, char **argv) {
	int last_error = 0;

	Get_GFX_Pointers(&pParam, &pBase, &pTexture, &pStereo);

	pParam->Linear = 1;
	pParam->T_layers = 0;
	pParam->Zoom = 1;
	pParam->AA = 4;
	pParam->AAr = 0;
	pParam->Startx = 0;
	pParam->Starty = 0;
	pParam->Front = 1.0;
	pParam->Distance = 0.6;
	pParam->Eyeshift = 0.0;

	pParam->T_level = T_BACK_LEVEL;
	pParam->Invert = 0;
	pParam->Triangles = 0;

	pBase->Data = NULL;
	pTexture->Data = NULL;
	pStereo->Data = NULL;
	base_file_name = NULL;
	texture_file_name = NULL;
	stereo_file_name = NULL;

	verbose = 0;

	if(parse_args(argc, argv))
		return -1;

	if(!stereo_file_name && base_file_name && texture_file_name && verbose) {
		fprintf(stderr, "verbose disabled for standard output\n");
		verbose = 0;
	}

	if(verbose) printf("Stereograph for linux Version %s\n", STEREOGRAPH_VERSION);

	if(base_file_name && (texture_file_name || (texture_width_to_use > 0))) {
		if(pParam->T_layers <= 0) {
			/* solid (single) tracing */
			if(verbose) printf("loading base '%s'...", base_file_name);
			if(!Read_Gfx_File(base_file_name, pBase)) {
				if(!texture_file_name) {
					if(verbose) printf("generating random texture (%i*%i)...", texture_width_to_use, pBase->Height);
					if(texture_width_to_use < pBase->Width) {
						last_error = Generate_Random_Texture(pTexture, texture_width_to_use, pBase->Height, random_texture_type);
					} else {
						if(verbose) printf("FAILED\n");
						fprintf(stderr, "texture_width_to_use is too big. This may not be greater than the base width.\n");
						last_error = -2;
					}
				} else {
					if(verbose) printf("loading texture '%s'...", texture_file_name);
					last_error = Read_Gfx_File(texture_file_name, pTexture);
					if(!last_error && (texture_width_to_use > 0)) {
						if(pTexture->Width >= texture_width_to_use) {
							last_error = Resize_GFX(pTexture, texture_width_to_use);
						} else {
							fprintf(stderr, "parametre error: texture_width_to_use is too big;\nmust be less than the texture width.\n");
							last_error = -2;
						}
					}
				}
				if(!last_error && pParam->Invert) {
					if((last_error = Invert_GFX(pBase))) {
						if(verbose) printf("FAILED\n");
						fprintf(stderr, "error inverting base;\n");
					}
				}
				if(!last_error) {
					/* setting up renderer; */
					if(verbose) printf("initializing renderer...");
					if(!(last_error = Initialize_Renderer())) {
						int y, s;
						if(verbose) printf("success;\n");
						if(verbose) printf("using following parametres (zadexy): %i %i %f %f %i %i;\n", pParam->Zoom, pParam->AA, pParam->Distance, pParam->Eyeshift, pParam->Startx, pParam->Starty);
						if(verbose && pParam->AAr)
							printf("anti-artefacts feature enabled;\n");
						if(verbose && !pParam->Linear)
							printf("linear rendering disabled;\n");
						if(verbose && pParam->Triangles)
							printf("triangle aid enabled, a pair will be added after rendering;\n");
						if(verbose) printf("processing...");
                                	
						for(y = 0, s = 0; (y < pBase->Height); y++) {
							ProcessLine(y);
							/* print simple status points (without refresh) */
							for( ; verbose && (s < (58 * y / pBase->Height)); s++)
								printf(".");
						}
						if(verbose) printf("completed;\n");

						if(pParam->Triangles) {
							if(verbose)
								printf("adding triangles to the final image...");
							Add_Triangles(pStereo->Width/2, 20*pParam->Zoom, pTexture->Width*pParam->Zoom, pStereo);
							if(verbose)
								printf("done;\n");
						}
	
						if(verbose) printf("writing stereogram to disk...");
						if(!Write_Gfx_File(stereo_file_name, output_format, pStereo) && verbose)
							printf("completed;\n");
					} else {
						if(verbose)
							printf("FAILED;\n");
						else    	
							fprintf(stderr, "initializing renderer...FAILED;\n");
						switch (last_error) {
							case -1 :
								fprintf(stderr, "illegal parametre;\n\n");
								print_info();
								break;
							case -2 :
								fprintf(stderr, "width of texture file greater than the base;\nthis state is not processable;\ntry the -w option to resize the texture.\n");
								break;
							case -3 :
								fprintf(stderr, "could not allocate memory.\n");
						}
						return -1;
					}
					Clear_Renderer();
				}
				free(pTexture->Data);
			}
			free(pBase->Data);
			/* end of single trace */
		} else {
			/* transparent trace */
			int z, a;
			
			Get_T_GFX_Pointers(&T_pBase, &T_pTexture);
			/* allocating multiple gfx data headers */
				(*T_pBase) = (struct GFX_DATA**) malloc(sizeof(struct GFX_DATA*) * (pParam->T_layers+1));
				(*T_pTexture) = (struct GFX_DATA**) malloc(sizeof(struct GFX_DATA*) * (pParam->T_layers+1));
				a = 2;
				if((*T_pBase) && (*T_pTexture)) {
					a = 0;
					for(z = 0; (z <= pParam->T_layers) && !a; z++) {
						*((*T_pBase)+z) = (struct GFX_DATA*) malloc(sizeof(struct GFX_DATA));
						(*((*T_pBase)+z))->Data = NULL;
						*((*T_pTexture)+z) = (struct GFX_DATA*) malloc(sizeof(struct GFX_DATA));
						(*((*T_pTexture)+z))->Data = NULL;
						if(!*((*T_pBase)+z) || !*((*T_pTexture)+z))
							a++;
					}
				}
				if(!a) {
					if(verbose) printf("rendering transparent layers enabled\nloading bases:\n");
					for(z = 0; (z <= pParam->T_layers) && !a; z++) {
						if(verbose) printf("%s ", T_base_file_name[z]);
						a = Read_Gfx_File(T_base_file_name[z], *((*T_pBase)+z));
					}
					if(!a) {
						if(verbose) printf("done;\nloading textures:\n");
						for(z = 0; (z <= pParam->T_layers) && !a; z++) {
							if(verbose) printf("%s ", T_texture_file_name[z]);
							a = Read_Gfx_File(T_texture_file_name[z], *((*T_pTexture)+z));
							if((texture_width_to_use > 0) && !a) {
								if((*((*T_pTexture)+z))->Width >= texture_width_to_use) {
									a = Resize_GFX((*((*T_pTexture)+z)), texture_width_to_use);
								} else {
									fprintf(stderr, "parametre error: texture_width_to_use is too big;\nmust be less than the texture width.\n");
									a = -2;
								}
							}
						}
						if(verbose && !a) printf("done;\n");
						if(!a && pParam->T_level) {
							a = T_adjust_level(*T_pBase, pParam->T_layers, pParam->T_level);
						}
						if(!a) {
							/* setting up renderer; */
							if(verbose) printf("initializing renderer...");
							if(!(last_error = Initialize_Renderer())) {
								int y, s;
								if(verbose) printf("success;\n");
								if(verbose) printf("using following parametres (zadexy): %i %i %f %f %i %i;\n", pParam->Zoom, pParam->AA, pParam->Distance, pParam->Eyeshift, pParam->Startx, pParam->Starty);
								if(verbose && pParam->AAr)
									printf("anti-artefacts feature enabled;\n");
								if(verbose) printf("processing...");
                                	
								for(y = 0, s = 0; (y < (**T_pBase)->Height); y++) {
									ProcessLine(y);
									/* print simple status points (without refresh) */
									for( ; verbose && (s < (58 * y / (**T_pBase)->Height)); s++)
										printf(".");
								}
              		                	
								if(verbose) printf("completed;\n");
	
								if(pParam->Triangles) {
									if(verbose)
										printf("adding triangles to the final image...");
									Add_Triangles(pStereo->Width/2, 20*pParam->Zoom, pTexture->Width*pParam->Zoom, pStereo);
									if(verbose)
										printf("done;\n");
								}
	
								if(verbose) printf("writing stereogram to disk...");
								if(!Write_Gfx_File(stereo_file_name, output_format, pStereo) && verbose)
									printf("completed;\n");
							} else {
								if(verbose)
									printf("FAILED;\n");
								else    	
									fprintf(stderr, "initializing renderer...FAILED;\n");
								switch (last_error) {
									case -1 :
										fprintf(stderr, "illegal parametre;\n\n");
										print_info();
										break;
									case -2 :
										fprintf(stderr, "width of the texture greater than the base;\nthis state is not processable;\ntry the -w option to resize the texture.\n");
										break;
									case -3 :
										fprintf(stderr, "could not allocate memory;\n");
										break;
									case -4 :
										fprintf(stderr, "base or texture image dimensions may NOT differ;\n");
										break;
									default :
										fprintf(stderr, "illegal parametre/unable to initialize renderer;\n\n");
										print_info();
								}
								return -1;
							}
							Clear_Renderer();
						}
						for(z = 0; z <= pParam->T_layers; z++)
							free((*((*T_pTexture)+z))->Data);
					}
					for(z = 0; (z <= pParam->T_layers) && (a < 2); z++) {
						free((*((*T_pBase)+z))->Data);
						free(*((*T_pTexture)+z));
						free(*((*T_pBase)+z));
					}
				} else {
					/* malloc for the T_pBase or T_pTexture indices failed */
					fprintf(stderr, "could not allocate memory for the gfx input headers\n");
					return -1;
				}
				free(*T_pBase);
				free(*T_pTexture);
			/* enf of transparent trace */
		}
	} else {
		if(!verbose) printf("Stereograph for linux Version %s\n", STEREOGRAPH_VERSION);
		fprintf(stderr, "NOTIFY: must specify at least a base and a texture file\nor define a texture width for a random texture.\nuse the -w option to define a texture width.\nnote that you CANNOT use random textures for transparent rendering.\n\n");
		print_info();
		return -1;
	}
	return 0;
}

int parse_args (int argc, char **argv) {
	int z, s = 1;
	
	for(z = 1; z < argc; z++, s = 1)
		while((z < argc) && ((argv[z][0] == '-') && (argv[z][s] != '\0')))
			switch(argv[z][s]) {
				case 'b' :
					if (z < (argc - 1)) {
						if(pParam->T_layers) {
							T_base_file_name = argv + ++z;
							z += pParam->T_layers;
						}
						base_file_name = argv[++z];
					} else {
						fprintf(stderr, "invalid argument '%c';\n", argv[z][s]);
						return -1;
					}
					break;
				case 't' :
					if (z < (argc - 1)) {
						if(pParam->T_layers) {
							T_texture_file_name = argv + ++z;
							z += pParam->T_layers;
						}
						texture_file_name = argv[++z];
					} else {
						fprintf(stderr, "invalid argument '%c';\n", argv[z][s]);
						return -1;
					}
					break;
				case 'o' :
					if (z < (argc - 1)) stereo_file_name = argv[++z]; else { fprintf(stderr, "invalid argument '%c';\n", argv[z][s]);  return -1; }
					break;
				/* the following is OBSOLETE and temporary left also under 'F' here for compatibility purpose only */
				case 'F' :
					if (z < (argc - 1)) {
						if(!strcmp(argv[++z], "png"))
							output_format = GFX_PNG;
						else if(!strcmp(argv[z], "ppm"))
							output_format = GFX_PPM;
						else if(!strcmp(argv[z], "tga"))
							output_format = GFX_TARGA;
						else
							output_format = atoi(argv[z]);
					} else { fprintf(stderr, "invalid argument '%c';\n", argv[z][s]);  return -1; }
					break;
				case 'f' :
					if (z < (argc - 1)) {
						if(!strcmp(argv[++z], "png"))
							output_format = GFX_PNG;
						else if(!strcmp(argv[z], "ppm"))
							output_format = GFX_PPM;
						else if(!strcmp(argv[z], "tga"))
							output_format = GFX_TARGA;
						else
							output_format = atoi(argv[z]);
					} else { fprintf(stderr, "invalid argument '%c';\n", argv[z][s]);  return -1; }
					break;
				case 'x' :
					if (z < (argc - 1)) pParam->Startx = atoi(argv[++z]); else { fprintf(stderr, "invalid argument '%c';\n", argv[z][s]);  return -1; }
					break;
				case 'y' :
					if (z < (argc - 1)) pParam->Starty = atoi(argv[++z]); else { fprintf(stderr, "invalid argument '%c';\n", argv[z][s]);  return -1; }
					break;
				case 'd' :
					if (z < (argc - 1)) pParam->Distance = (float) atof(argv[++z]); else { fprintf(stderr, "invalid argument '%c';\n", argv[z][s]);  return -1; }
					break;
				case 'e' :
					if (z < (argc - 1)) pParam->Eyeshift = (float) atof(argv[++z]); else { fprintf(stderr, "invalid argument '%c';\n", argv[z][s]);  return -1; }
					break;
				case 'a' :
					if (z < (argc - 1)) pParam->AA = atoi(argv[++z]); else { fprintf(stderr, "invalid argument '%c';\n", argv[z][s]);  return -1; }
					break;
				case 'z' :
					if (z < (argc - 1)) pParam->Zoom = atoi(argv[++z]); else { fprintf(stderr, "invalid argument '%c';\n", argv[z][s]);  return -1; }
					break;
				case 'w' :
					if (z < (argc - 1)) texture_width_to_use = atoi(argv[++z]); else { fprintf(stderr, "invalid argument '%c';\n", argv[z][s]);  return -1; }
					break;
				case 'l' :
					if (z < (argc - 1)) {
						if(!strcmp(argv[++z], "top"))
							pParam->T_level = T_TOP_LEVEL;
						else if(!strcmp(argv[z], "back"))
							pParam->T_level = T_BACK_LEVEL;
						else if(!strcmp(argv[z], "none"))
							pParam->T_level = T_NO_LEVEL;
						else
							pParam->T_level = atoi(argv[z]);
					} else {
						fprintf(stderr, "invalid argument '%c';\n", argv[z][s]);  return -1;
					}
					break;
				case 'I' :
					pParam->Invert = 1;
					s++;
					break;
				case 'C' :
					random_texture_type = TEX_COLORED;
					s++;
					break;
				case 'G' :
					random_texture_type = TEX_GRAYSCALE;
					s++;
					break;
				case 'M' :
					random_texture_type = TEX_BW;
					s++;
					break;
				case 'S' :
					random_texture_type = TEX_SINLINE;
					s++;
					break;
				case 'R' :
					pParam->AAr = 1;
					s++;
					break;
				case 'L' :
					pParam->Linear = 0;
					s++;
					break;
				/* the following two parametres are obsolete and only for compatibility temporary intact */
				/*case 'r' :
					pParam->AAr = 1;
					s++;
					break;
				case 'l' :
					pParam->Linear = 0;
					s++;
					break;*/
				case 'A' :
					pParam->Triangles = 1;
					s++;
					break;
				case 'v' :
					verbose = 1;
					s++;
					break;
				case 'T' :
					if(z > 1) { fprintf(stderr, "invalid argument '%c';\nthe T descriptor must PRECEED any argument!\n", argv[z][s]);  return -1; }
					if (z < (argc - 1)) pParam->T_layers = atoi(argv[++z]); else { fprintf(stderr, "invalid argument '%c';\n", argv[z][s]);  return -1; }
					break;
				case 'V' :
					printf("stereograph-%s\n", STEREOGRAPH_VERSION);
					return -2;
				default:
					fprintf(stderr, "invalid argument '%c';\n", argv[z][s]);
					return -1;
			}
	return 0;
}
